sql_metadata/parser.py

```diff
@@ -107,12 +129,14 @@ def tokens(self) -> List[SQLToken]:
             elif token.is_right_parenthesis:
                 self._determine_closing_parenthesis_type(token=token)
 
-            if tok.is_keyword and tok.normalized not in KEYWORDS_IGNORED:
+            if tok.is_keyword and "".join(tok.normalized.split()) in RELEVANT_KEYWORDS:
                 last_keyword = tok.normalized
             token.is_in_nested_function = self._is_in_nested_function
+            token.parenthesis_level = self._parenthesis_level
             tokens.append(token)
 
         self._tokens = tokens
+        _ = self.query_type
```

Comment:
Why are we "pre-caching" this property here?

Reply:
Cause `tokens` are used in all other (or are like begging of the chain), and `query_type` raises exception if not supported query is used to make sure you won't get weird results for not supported queries.

Can be moved to init probably but it requires the tokens to be set so init will take longer (but then again tokens are used in each step (directly or through chain) and they are cached so it can be also checked in `__init__`

Reply:
Ah I know! I didn't put this in the init as tokens are not used in generalization so that would affect the performance for those (just remembered 😅)

Reply:
Makes sense. However, can't we simply "calculate" the query type when user asks for it, i.e lazy calculate it?

Reply:
We can, but than you need a check for `query_type` in each property/method to raise the exception on unsupported queries. 

I don't know if we need this exception or not, my idea behind this was that users tend to report bugs while in fact they are not bugs but enhancements, so if user tries to parse some weird query that we do not support we make it clear for him that it's not supported (on the other hand maybe that query would have worked ok 🤷 ). That way we can also add tests for new type of queries we do not cover now (even if they work fine and no one reported a bug).

If we remove the exception it can be left only as a property (and then it's used only to check if query_type is create in `columns` and `tables`) and calculated when needed. (on the other hand it's like super simple so should not affect the performance anyway)

Reply:
Now it makes sense. Can you then add a comment saying why this line is there?

Reply:
Added comment